enhancements/cvo/clusteroperator-resource-handling.md

```diff
@@ -0,0 +1,117 @@
+an operator pod runs.
+This is critical for debugging clusters that fail to install or fail to upgrade with new operators present.
+
+## Motivation
+
+Debugging information for clusters that succeed in bootstrapping, but fail during installation is missing most
+of the data required to resolve the issue via must-gather.
+
+### Goals
+
+1. allow collection of debugging data for failed installs using normal tools. 
+
+### Non-Goals
+
+1. create a new tool to gather data for failed installs after bootstrapping.
+2. taking responsibility for creating clusteroperator resources.  Individual operators are responsible for creating and maintaining
+   clusteroperator resources.
+
+## Proposal
+
+1. `clusteroperator` resources in the payload should be created with the required status conditions (available, progressing,
+   degraded) set to `Unknown`.
+2. `clusteroperator` creation by the CVO needs to honor or update `.status.relatedResources`.  This requires updating
+    status after the creation.
+3.  `clusteroperator` resources in the payload should all be created immediately regardless of where in the payload ordering
+    they are located.  This ensures that they are always present during collection.
+4.  The CVO waiting logic on `clusteroperator` remains the same.
+
+### Specific Implementation Option
+
+This isn't a required mechanism for implementation, but it demonstrates how narrowly scoped the change is.
+ 1. create a new control loop with a clusteroperator lister, clusteroperator client, and a function to get the current payload.
+ 2. register event handlers on clusteroperator informer and time based every minute.
+ 3. sync loop reads the current payload.  for each clusteroperator in the payload
+    1. check lister to see if clusteroperator exists.  If so, continue to next clusteroperator.
+    2. create clusteroperator with empty spec and metadata.  If create fails, continue to next clusteroperator.
+    3. update clusteroperator/status with `.status.relatedResources` and the three required conditions in `Unknown` state.
+
+There is no need to modify the existing CVO logic because it's all valid.
+The individual operators are controllers so the presence of an unknown state doesn't matter.
+The new control loop doesn't fight with any individual operators because it's a create-only call with a one time status priming.
+If anything goes wrong with this control loop, the rest of the system continues to function as it does today. 
+
+### Risks and Mitigations
+
+1. Existing clusteroperators may treat presence and absence or condition==Unknown as special and fail to reconcile.
+   This would be a bug in the operator implementation that needs to be fixed.
+
+## Design Details
+
+### Test Plan
+
+1. When an install in CI fails at some point in the release, we should see must-gather information
+2. During an installation, the `clusteroperator` resources should be visible via the API immediately.
```

Comment:
CI coverage for this `immediately` seems tricky.  Were you expecting to use audit logs or some such?

Reply:
> CI coverage for this `immediately` seems tricky. Were you expecting to use audit logs or some such?

I'm confident enough that I'll investigate a "cluster didn't install" problem next release that I'll know if it works.

Reply:
> I'm confident enough that I'll investigate a "cluster didn't install" problem next release that I'll know if it works.

That's 1 though.  I don't see how 2 is all that important on its own, certainly not enough to be worth specific CI coverage.  I'm happy if 1 gets CI coverage.  I'm not happy if something important enough to be in the Test Plan is covered by "@deads2k manually looks into this occasionally" ;).

Reply:
> > I'm confident enough that I'll investigate a "cluster didn't install" problem next release that I'll know if it works.
> 
> That's 1 though. I don't see how 2 is all that important on its own, certainly not enough to be worth specific CI coverage. I'm happy if 1 gets CI coverage. I'm not happy if something important enough to be in the Test Plan is covered by "@deads2k manually looks into this occasionally" ;).

Presumably a unit test will work.